### DISCUSSION TITLE
Pester inheritance

### DISCUSSION THREAD

**Discussion by guytwig** [pester/Pester#2000] [Q&A] [answered]
Hi,
Is there a way to inherit\extend test cases?

More detailed:
I have a TestBase class which have common functionalities and modules for all the tests that inherit it.
Each inherited test calls “Run” function. The run function (implemented in BaseTest) is calling the test steps: Init, Configure, Run, Analyze and finally Teardown. Each test decide weather to override the
step or not.
I would like to have one .tests file that will get the name of the test and will run its steps.

To solve it I implemented a Factory module that gets the test name and returns the test object.
This object has the same functionalities that BaseTest has so I can have the one .tests file with the steps as test cases.

In general I would like to move BaseTest implementation to a .tests file and inherit it (Is it the correct way?).
What’s the best way to use Pester in that case?

**Comment by nohwnd** [comment]
Do you a have an example with more details (but hopefully small example without too many details)?

**Reply by JustinGrote** [reply]
I usually use a Shared.ps1 or Shared.psm1 that I load in some complicated test functions at times so my actual tests can read fairly cleanly.

**Comment by nohwnd** [comment, marked as answer]
Pester has no inheritance, there are no test classes to derive from. The blocks are just scriptblocks that run.

What you can do is put the shared code in a helper module and import it in `BeforeAll`, and use `BeforeAll` / `BeforeEach` on a parent `Describe` or `Context` so nested tests inherit the setup. For running the same steps over many inputs use `-ForEach` on `Describe` / `Context` / `It`, or `New-PesterContainer -Data`.
